src/modules/pair/services/pair.metadata.builder.ts

```diff
@@ -36,126 +57,266 @@ export class PairsMetadataBuilder {
+
+        this.pairs = PairFilteringService.pairsByLpTokenIds(
             this.filters,
-            this.pairsMetadata,
+            this.pairs,
         );
         return this;
     }
 
     async filterByFarmTokens(): Promise<PairsMetadataBuilder> {
-        this.pairsMetadata = await this.filteringService.pairsByFarmTokens(
+        if (
+            !(this.filters instanceof PairsFilter) ||
+            !this.filters.farmTokens ||
+            this.filters.farmTokens.length === 0
+        ) {
+            return this;
+        }
+
+        const farmTokens = await Promise.all(
+            this.pairs.map((pairMetadata) =>
+                this.pairCompute.getPairFarmToken(pairMetadata.address),
+            ),
+        );
+
+        this.pairs = PairFilteringService.pairsByFarmTokens(
             this.filters,
-            this.pairsMetadata,
+            this.pairs,
+            farmTokens,
         );
         return this;
     }
 
     async filterByState(): Promise<PairsMetadataBuilder> {
-        this.pairsMetadata = await this.filteringService.pairsByState(
+        const pairsStates = await this.pairService.getAllStates(
+            this.pairs.map((pair) => pair.address),
+        );
+
+        this.pairs.forEach((pair, index) => (pair.state = pairsStates[index]));
+
+        this.pairs = PairFilteringService.pairsByState(
             this.filters,
-            this.pairsMetadata,
+            this.pairs,
         );
         return this;
     }
 
     async filterByFeeState(): Promise<PairsMetadataBuilder> {
-        this.pairsMetadata = await this.filteringService.pairsByFeeState(
+        const pairsFeeStates = await this.pairService.getAllFeeStates(
+            this.pairs.map((pair) => pair.address),
+        );
+
+        this.pairs.forEach(
+            (pair, index) => (pair.feeState = pairsFeeStates[index]),
+        );
+
+        this.pairs = PairFilteringService.pairsByFeeState(
             this.filters,
-            this.pairsMetadata,
+            this.pairs,
         );
         return this;
     }
 
     async filterByVolume(): Promise<PairsMetadataBuilder> {
-        this.pairsMetadata = await this.filteringService.pairsByVolume(
+        const pairsVolumes = await this.pairCompute.getAllVolumeUSD(
+            this.pairs.map((pair) => pair.address),
+        );
+
+        this.pairs.forEach(
+            (pair, index) => (pair.volumeUSD24h = pairsVolumes[index]),
+        );
+
+        this.pairs = PairFilteringService.pairsByVolume(
             this.filters,
-            this.pairsMetadata,
+            this.pairs,
         );
         return this;
     }
 
     async filterByLockedValueUSD(): Promise<PairsMetadataBuilder> {
-        this.pairsMetadata = await this.filteringService.pairsByLockedValueUSD(
+        const pairsLiquidityUSD = await this.pairService.getAllLockedValueUSD(
+            this.pairs.map((pair) => pair.address),
+        );
+
+        this.pairs.forEach(
```

Comment:
it wouldn't make a much difference here, but for a high number of iterations, `forEach` is the least efficient way of ranging:
- https://dev.to/maafaishal/benchmarking-for-while-forof-and-arrayforeach-using-performancenow-1jjg